File: nodejs-typescript/src/domain/AccountNumber.ts
import { InvalidAccountNumberError } from './errors.js';

const ACCOUNT_NUMBER_PATTERN = /^\d{16}$/;

export class AccountNumber {
  private constructor(private readonly value: string) {}

  static parse(value: string): AccountNumber {
    const trimmed = value.trim();
    if (trimmed.length === 0) {
      throw new InvalidAccountNumberError('Account number cannot be blank');
    }
    if (!/^\d+$/.test(trimmed)) {
      throw new InvalidAccountNumberError('Account number must contain only digits');
    }
    if (!ACCOUNT_NUMBER_PATTERN.test(trimmed)) {
      throw new InvalidAccountNumberError('Account number must be exactly 16 digits');
    }
    return new AccountNumber(trimmed);
  }

  toString(): string {
    return this.value;
  }

  equals(other: AccountNumber): boolean {
    return this.value === other.value;
  }
}

File: dotnet/src/MableBank.Domain/AccountNumber.cs
using System.Text.RegularExpressions;
using MableBank.Domain.Errors;

namespace MableBank.Domain;

public readonly partial record struct AccountNumber
{
    [GeneratedRegex(@"^\d{16}$", RegexOptions.CultureInvariant)]
    private static partial Regex SixteenDigitsPattern();

    public string Value { get; }

    private AccountNumber(string value)
    {
        Value = value;
    }

    public static AccountNumber Parse(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new InvalidAccountNumberException("Account number cannot be blank");
        }

        var trimmed = value.Trim();
        if (!trimmed.All(char.IsDigit))
        {
            throw new InvalidAccountNumberException("Account number must contain only digits");
        }

        if (!SixteenDigitsPattern().IsMatch(trimmed))
        {
            throw new InvalidAccountNumberException("Account number must be exactly 16 digits");
        }

        return new AccountNumber(trimmed);
    }

    public override string ToString() => Value;
}
